### PR TITLE
basic graphite usage

### PR DESCRIPTION
**Pull request by evemorgen**
- class for basic `graphite` calls,
 - introducing config gem

### CODE REVIEW

File: rails_app/app/adapters/graphite.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+require 'net/http'
+require 'json'
+require 'graphite-api'
+
+class GraphiteAdapter
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: rename file to graphite_adapter

File: rails_app/app/adapters/graphite.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        @graphite_host = Settings.graphite.host
+        @default_period = Settings.graphite.default_get_metric_time
+        @api = GraphiteAPI.new(
+            graphite: "#{@graphite_host}:#{Settings.graphite.port}",
+            prefix: Settings.graphite.metric_prefix
+        )
+    end
+
+    def send_raw(metric, value = nil, time = nil)
+        if !metric.is_a?(Hash)
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: ```metric = { metric => value } unless metric.is_a?(Hash)```

File: rails_app/app/adapters/graphite.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            metric = { metric => value }
+        end
+
+        if time
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: ```time.present? ? @api.metrics(metric, Time.at(time)) : @api.metrics(metric)```

File: rails_app/app/adapters/graphite.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                format: type,
+                from: time_from ? time_from : @default_period,
+                to: time_to ? time_to : "now"
+            }.to_query
+        })
+        req = Net::HTTP::Get.new(url.to_s)
+        res = Net::HTTP.start(url.host, url.port) {|http|
+            http.request(req)
+        }
+        return JSON.parse(res.body)[0]["datapoints"]
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: redundant return, single quotes 'datapoints'

File: rails_app/app/adapters/graphite.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    def get_metric_json(metric, time_from: nil, time_to: nil)
+        return get_metric(metric, "json", time_from: time_from, time_to: time_to)
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: redundant return, single quotes 'json' (double for interpolation)

File: rails_app/app/adapters/graphite.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    def get_metric_png(metric, time_from: nil, time_to: nil)
+        return get_metric(metric, "png", time_from: time_from, time_to: time_to)
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: redundant return, single quotes 'png' (double for interpolation)

File: rails_app/app/adapters/graphite.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
+require 'net/http'
+require 'json'
+require 'graphite-api'
+
+class GraphiteAdapter
+    def initialize
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: yo use two spaces for indentation

File: rails_app/app/adapters/graphite.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    
+    def get_metric(metric, type, time_from: nil, time_to: nil)
+        url = URI::HTTP.build({
+            host: @graphite_host,
+            path: "/render",
+            query: {
+                target: metric,
+                format: type,
+                from: time_from ? time_from : @default_period,
+                to: time_to ? time_to : "now"
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: havent u meant Time.now insted "now" ?
if not single quotes 'now'

File: rails_app/app/adapters/graphite.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        end
+
+        if time
+            @api.metrics(metric, Time.at(time))
+        else
+            @api.metrics(metric)
+        end
+    end
+
+    def send_deal_price(deal_name, price, time = nil, user = 'internal')
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: never used time?

File: rails_app/app/adapters/graphite.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        )
+    end
+
+    def send_raw(metric, value = nil, time = nil)
+        if !metric.is_a?(Hash)
+            metric = { metric => value }
+        end
+
+        if time
+            @api.metrics(metric, Time.at(time))
```
Comment (szmo) [inline rails_app/app/adapters/graphite.rb]: time gonna be seconds since unix epoch or its mistake?

**Review by szmo — CHANGES_REQUESTED** [review]

File: rails_app/app/adapters/graphite.rb (hunk 6 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: `When it runs out of code or "reaches the end", in which case the value of the last expression is returned` - got it

File: rails_app/app/adapters/graphite.rb (hunk 5 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: done

File: rails_app/app/adapters/graphite.rb (hunk 4 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: yep, done

File: rails_app/app/adapters/graphite.rb (hunk 8 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: nope, I ment `now` as a string, graphite can operate on this type of syntax:
http://graphite.readthedocs.io/en/latest/render_api.html#from-until

File: rails_app/app/adapters/graphite.rb (hunk 9 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: forgot to pass as a parameter to `send_raw`, thanks!

File: rails_app/app/adapters/graphite.rb (hunk 10 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: yep

File: rails_app/app/adapters/graphite.rb (hunk 8 above)
Comment (szmo) [reply rails_app/app/adapters/graphite.rb]: kk

File: rails_app/app/adapters/graphite.rb (hunk 7 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: It makes me feel like javascript, ugh!
done

File: rails_app/app/adapters/graphite.rb (hunk 2 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: I think this is counter intuitive to read that way?

File: rails_app/app/adapters/graphite.rb (hunk 3 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: the hell is this `.present?` thing?
> An object is present if it’s not blank.

`.nil?` seems like a better choice

File: rails_app/app/adapters/graphite.rb (hunk 1 above)
Comment (evemorgen) [reply rails_app/app/adapters/graphite.rb]: ok, done

File: rails_app/app/adapters/graphite.rb (hunk 3 above)
Comment (szmo) [reply rails_app/app/adapters/graphite.rb]: could be

**Comment by szmo** [comment]
and please add ```config.autoload_paths += %W[#{config.root}/app/adapters/)]``` to ```config/application.rb``` inside application class, then our GraphiteAdapter will be autoload and in any place in code will be possible to use GraphiteAdapter

**Comment by evemorgen** [comment]
rails are doing that automatically:
```
21:07:33 evemorgen@mac.home rails_app feature/graphite_api ? rails r 'puts ActiveSupport::Dependencies.autoload_paths'                            130 ↵
Running via Spring preloader in process 30364
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/adapters
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/assets
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/channels
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/controllers
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/controllers/concerns
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/helpers
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/jobs
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/mailers
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/models
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/app/models/concerns
/Users/evemorgen/Dropbox/projekty/price_monitor/rails_app/spec/mailers/previews

```

**Review by szmo — DISMISSED** [review]

**Review by szmo — APPROVED** [review]

**Pull request merged**